mirascope/base/prompts.py

```diff
@@ -0,0 +1,184 @@
+"""A base class for writing prompts."""
+import re
+from string import Formatter
+from textwrap import dedent
+from typing import Any, Callable, ClassVar, Type, TypeVar, Union
+
+from pydantic import BaseModel
+
+from .types import (
+    AssistantMessage,
+    Message,
+    ModelMessage,
+    SystemMessage,
+    ToolMessage,
+    UserMessage,
+)
+
+
+class BasePrompt(BaseModel):
+    '''The base class for working with prompts.
+
+    This class is implemented as the base for all prompting needs across various model
+    providers.
+
+    Example:
+
+    ```python
+    from mirascope import BasePrompt
+
+
+    class BookRecommendationPrompt(BasePrompt):
+        """A prompt for recommending a book."""
+
+        prompt_template = """\\
+            SYSTEM: You are the world's greatest librarian.
+            USER: Please recommend a {genre} book.
+        """
+
+        genre: str
+
+
+    prompt = BookRecommendationPrompt(genre="fantasy")
+    print(prompt.messages())
+    #> [{"role": "user", "content": "Please recommend a fantasy book."}]
+
+    print(prompt)
+    #> Please recommend a fantasy book.
+    ```
+    '''
+
+    tags: ClassVar[list[str]] = []
+    prompt_template: ClassVar[str] = ""
+
+    def __str__(self) -> str:
+        """Returns the formatted template."""
+        return self._format_template(self.prompt_template)
+
+    def messages(self) -> Union[list[Message], Any]:
+        """Returns the template as a formatted list of messages."""
+        message_type_by_role = {
+            "system": SystemMessage,
```

Comment:
create enum for system, user assistant, model, tool. then use enum as keys here. also in gemini/calls.py, openai/calls.py, and anthropic/calls.py

Reply:
Done.